### PR TITLE
[P2P] KISS 2 - Peer discovery [Merge me after #520] - (Issues: #416, #429)

### LINKED ISSUE #416: [P2P] Implement KISS peer discovery for LocalNet
## Objective

Ensure that the `debug client` is capable of broadcasting / sending messages to the `LocalNet` nodes even after the `ValidatorMap` (which contained the addresses of the nodes, sourced from the genesis) has been sunset.

## Origin Document

Since #203 and #331 replace `ValidatorMap` with dynamic actor fetching from the persistence layer based on the Node's current height, the `debug client` needs a mechanism to discover the entire local set of peers regardless of the address book each individual node has.

This leads us to the first real use-case for **peer-discovery**.

## Goals

- Design and implement a way for gathering the current (i.e. at the current height) validator set of addresses from the point of view of the debug client or another node that intends to join the network.

## Deliverable

- [ ] Functioning `LocalNet` without a hardcoded set of `validatorAddresses`
- [ ] Updated RPC to exposes a new endpoint that returns the addresses of the staked validators known to the node the client is connected to
- [ ] Bonus: allow a graphical and/or CLI based way for inspecting the addressBook for development and debugging purposes

## Non-goals / Non-deliverables

- 

## General issue deliverables

- [ ] Update the appropriate CHANGELOG(s)
- [ ] Update any relevant local/global README(s)
- [ ] Update relevant source code tree explanations
- [ ] Add or update any relevant or supporting [mermaid](https://mermaid-js.github.io/mermaid/) diagrams

## Testing Methodology

- [ ] **New tests**: Documentation (e.g. video) showing the usage & functionality of the new RPC in action
- [ ] **All tests**: `make test_all`
- [ ] **LocalNet**: verify a `LocalNet` is still functioning correctly by following the instructions at [docs/development/README.md](https://github.com/pokt-network/pocket/tree/main/docs/development)

---

**Creator**: [@deblasis]
**Co-Owners**: [@Olshansk]

### LINKED ISSUE #429: [P2P] Refactor `CreateWithProvider` using the new `bus` based DI
## Objective

Improve code-health and consistency of patterns

## Origin Document

While developing #402 and #331 I had to inject some dependencies into the `P2P` module. This was our first use case in our codebase and the initial approach was indeed marked as a `HACK` from the beginning but there was something missing that is now available via #425: the ability to register modules dynamically on the `bus` instance, effectively enabling Dependency Injection

This ticket was descoped from #425 which became quite large and scope-creepy:

https://github.com/pokt-network/pocket/pull/425#discussion_r1062837932

## Goals

- Remove the `CreateWithProviders` constructor and use the standardized `Create` instead.
- Refactoring the way we access the dependencies `AddrBookProvider` and `CurrentHeightProvider` by registering them at the caller site and consuming them inside the `P2P` `Create()`

## Deliverable

- [ ] Refactored `P2P` module `Create`
- [ ] Dependency injection for `AddrBookProvider` and `CurrentHeightProvider`

## Non-goals / Non-deliverables

- Changes in `P2P` logic

## General issue deliverables

- [ ] Update the appropriate CHANGELOG(s)
- [ ] Update any relevant local/global README(s)
- [ ] Update relevant source code tree explanations
- [ ] Add or update any relevant or supporting [mermaid](https://mermaid-js.github.io/mermaid/) diagrams

## Testing Methodology

- [ ] **Task specific tests or benchmarks**: `make ...`
- [ ] **New tests or benchmarks**: `make ...`
- [ ] **All tests**: `make test_all`
- [ ] **LocalNet**: verify a `LocalNet` is still functioning correctly by following the instructions at [docs/development/README.md](https://github.com/pokt-network/pocket/tree/main/docs/development)

---

**Creator**: @deblasis
**Co-Owners**: @Olshansk

### PR DESCRIPTION
**Pull request by deblasis**
## Description

This PR has been extracted from https://github.com/pokt-network/pocket/pull/491 and is, hopefully, more digestible from a code-review and scope point of view.

The main goal is to remove hardcoded nodes and move towards a more dynamic environment.

It's also highlighting the potential entry points for subsequent P2P work

The code leverages the abstractions added recently (`currentHeightProvider` and `addressBookProvider`) to fetch the data from an RPC endpoint.

## Issue

Fixes #416
Fixes #429

## Type of change

Please mark the relevant option(s):

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [x] Major breaking change
- [ ] Documentation
- [ ] Other 

### CLI

- Updated CLI to use to source the address book and the current height from the RPC server leveraging the `rpcAddressBookProvider` and `rpcCurrentHeightProvider` respectively and the `bus` for dependency injection

### P2P

- Modules embed `base_modules.IntegratableModule` and `base_modules.InterruptableModule` for DRYness
- Deprecated `debugAddressBookProvider`
- Added `rpcAddressBookProvider` to source the address book from the RPC server
- Leveraging `bus` for dependency injection of the `addressBookProvider` and `currentHeightProvider`
- Deprecated `debugCurrentHeightProvider`
- Added `rpcCurrentHeightProvider` to source the current height from the RPC server
- Fixed raintree to use the `currentHeightProvider` instead of consensus (that was what we wanted to avoid in the first place)
- Added `getAddrBookDelta` to calculate changes to the address book between heights and update the internal state and componentry accordingly
- Reacting to `ConsensusNewHeightEventType` to update the address book 
- Updated tests

### RPC

- Updated RPC to expose the node's address book via GET /v1/p2p/staked_actors_address_book

## Testing

- [x] `make develop_test`
- [x] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [x] I have updated the corresponding CHANGELOG

### If Applicable Checklist

- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

File: app/client/cli/debug.go [hunk 1]
```diff
@@ -44,18 +47,27 @@ var (
...
-
 	configPath  string = runtime.GetEnv("CONFIG_PATH", "build/config/config1.json")
 	genesisPath string = runtime.GetEnv("GENESIS_PATH", "build/config/genesis.json")
+	rpcHost     string
 )
 
+// NOTE: this is required by the linter, otherwise a simple string constant would have been enough
+type cliContextKey string
+
+const busCLICtxKey = "bus"
```
Comment (deblasis) [inline app/client/cli/debug.go:58]: Re: https://github.com/pokt-network/pocket/pull/491/files#r1107988965

I simplified using context only to pass around the `bus` and then we use it as usual

Comment (Olshansk) [reply app/client/cli/debug.go:58]: 🚌

File: app/client/cli/utils.go [hunk 2]
```diff
@@ -187,3 +187,12 @@ func rpcResponseCodeUnhealthy(statusCode int, response []byte) error {
...
 	return fmt.Sprintf("\033[1m%s\033[0m", s)
 }
+
+func setValueInCLIContext(cmd *cobra.Command, key cliContextKey, value any) {
+	cmd.SetContext(context.WithValue(cmd.Context(), key, value))
+}
+
+func getValueFromCLIContext[T any](cmd *cobra.Command, key cliContextKey) (T, bool) {
+	value, ok := cmd.Context().Value(key).(T)
+	return value, ok
```
Comment (Olshansk) [inline app/client/cli/utils.go:197]:
```suggestion
	return cmd.Context().Value(key).(T)
```

File: rpc/v1/openapi.yaml [hunk 3]
```diff
@@ -1,6 +1,6 @@
 openapi: 3.0.3
```
Comment (Olshansk) [inline rpc/v1/openapi.yaml:1]: You need to run `generate_cli_commands_docs` so the docs are updated

File: rpc/handlers.go [hunk 4]
```diff
@@ -67,3 +69,77 @@ func (s *rpcServer) broadcastMessage(msgBz []byte) error {
...
+			protocolActorGetter = persistenceContext.GetAllApps
+		case Fisherman:
+			protocolActorGetter = persistenceContext.GetAllFishermen
+		case ServiceNode:
+			protocolActorGetter = persistenceContext.GetAllServiceNodes
+		case Validator:
+			protocolActorGetter = persistenceContext.GetAllValidators
+		}
+	}
+	return protocolActorGetter
```
Comment (Olshansk) [inline rpc/handlers.go:144]: Just some style
```suggestion
	if params.ActorType == nil {
		return persistenceContext.GetAllStakedActors
	}
	switch *params.ActorType {
	case Application:
		return persistenceContext.GetAllApps
	case Fisherman:
		return persistenceContext.GetAllFishermen
	case ServiceNode:
		return persistenceContext.GetAllServiceNodes
	case Validator:
		return persistenceContext.GetAllValidators
	}
```

File: p2p/providers/addrbook_provider/rpc/provider.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+	rpcUrl    string
+	p2pCfg    *configs.P2PConfig
+	rpcClient *rpc.ClientWithResponses
+
+	connFactory typesP2P.ConnectionFactory
+}
+
+func NewRPCAddrBookProvider(options ...modules.ModuleOption) *rpcAddrBookProvider {
+	dabp := &rpcAddrBookProvider{
+		rpcUrl:      fmt.Sprintf("http://%s:%s", rpcHost, defaults.DefaultRPCPort),
```
Comment (Olshansk) [inline p2p/providers/addrbook_provider/rpc/provider.go:45]:
```suggestion
		rpcUrl:      fmt.Sprintf("http://%s:%s", rpcHost, defaults.DefaultRPCPort), // TODO: Make port configurable
```

File: p2p/addrbook_delta_test.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+		after  typesP2P.AddrBook
+	}
+	tests := []struct {
+		name        string
+		args        args
+		wantAdded   []*typesP2P.NetworkPeer
+		wantRemoved []*typesP2P.NetworkPeer
+	}{
+		{
+			name: "empty slices should return empty slices",
```
Comment (Olshansk) [inline p2p/addrbook_delta_test.go:23]: No 🍕 here...

File: p2p/providers/addrbook_provider/rpc/provider.go [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+		connFactory: transport.CreateDialer, // default connection factory, overridable with WithConnectionFactory()
+	}
+
+	for _, o := range options {
+		o(dabp)
+	}
+
+	initRPCClient(dabp)
```
Comment (Olshansk) [inline p2p/providers/addrbook_provider/rpc/provider.go:53]: Make this more functional. One of these two:

1. `dabp.rpcClient = initRPCClient(dabp.rpcUrl)`
2. `dabp.initRPCClient()`

File: p2p/providers/current_height_provider/rpc/provider.go [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+func NewRPCCurrentHeightProvider(options ...modules.ModuleOption) *rpcCurrentHeightProvider {
+	dabp := &rpcCurrentHeightProvider{
+		rpcUrl: fmt.Sprintf("http://%s:%s", rpcHost, defaults.DefaultRPCPort),
+	}
+
+	for _, o := range options {
+		o(dabp)
+	}
+
+	initRPCClient(dabp)
```
Comment (Olshansk) [inline p2p/providers/current_height_provider/rpc/provider.go:75]: Ditto previous comment

**Review by Olshansk — APPROVED** [review]
Left a couple comments (nothing blocking) but PTAL before merging! Again, 🙏 for splitting this!!!!

File: app/client/cli/utils.go (hunk 2 above)
Comment (deblasis) [reply app/client/cli/utils.go]: It doesn't work. Applied and reverted

File: rpc/v1/openapi.yaml (hunk 3 above)
Comment (deblasis) [reply rpc/v1/openapi.yaml:1]: `generate_cli_commands_docs` updates docs for the CLI. (this has to be ran in the PR that adds the `non_interactive` flag if I didn't do that already. Noted.

In here I have already ran `make generate_rpc_openapi` before adding the handlers for the new endpoint

**Pull request merged**